query-algebrizer/src/clauses/mod.rs

```diff
@@ -806,26 +807,24 @@ impl ConjoiningClauses {
     }
 
 
-    /// When a CC has accumulated all patterns, generate value_type_tag entries in `wheres`
-    /// to refine value types for which two things are true:
-    ///
-    /// - There are two or more different types with the same SQLite representation. E.g.,
-    ///   ValueType::Boolean shares a representation with Integer and Ref.
-    /// - There is no attribute constraint present in the CC.
-    ///
-    /// It's possible at this point for the space of acceptable type tags to not intersect: e.g.,
-    /// for the query
-    ///
-    /// ```edn
-    /// [:find ?x :where
-    ///  [?x ?y true]
-    ///  [?z ?y ?x]]
-    /// ```
-    ///
-    /// where `?y` must simultaneously be a ref-typed attribute and a boolean-typed attribute. This
-    /// function deduces that and calls `self.mark_known_empty`. #293.
+    /// For each bound var. discover which ones have 
     pub fn expand_type_tags(&mut self) {
-        // TODO.
+        let mut vars_to_remove = Vec::new();
+        for (v, tag) in self.extracted_types.iter() {
+            if let Some(cols) = self.column_bindings.get(&v) {
```

Comment:
I don't think this is correct.

What we'd like to do is find SQL patterns for which the attribute place is insufficient to determine an unambiguous SQL representation of the value place, and impose additional restrictions.

What you've done here is found variables for which the type isn't known (which is better done by walking `column_bindings` and looking at `self.known_type_set(&var)` — `extracted_types` isn't reliable for this purpose), generating cross-links to other column-uses of the variable.

Sometimes those two things are the same, and sometimes they're not.

We don't need to generate tag comparisons if the potential types don't share a SQL representation. We sometimes _do_ need to generate type code constraints when the type of a place is later known precisely.

For example,

```
[:find …
 [?x :foo/bar ?y]
 [?z _ ?x]]
```

— `?x` is known for sure to have `ValueType::Ref`, but that query translates to

```
SELECT …
FROM datoms AS datoms00, all_datoms AS all_datoms01
WHERE datoms00.a = 99 AND datoms00.e = all_datoms00.v
```

which is incorrect. This is even true when there's no other pattern:

```
[:find …
 [?x _ ?y]
 [(< ?y 10)]]
```

should constrain `?y` to numeric values, type code `5`. (#385 would cover dropping back to `datoms` from `all_datoms` in this case, too.)

My suggestion is this:

As each pattern is applied, we know _within the scope of the particular contributing table_ which value types might be returned: `?e` `?a` and `?tx` will always be refs, and `?v` will, in the naive case, always be either `ValueTypeSet::one_of(range_of_a)` or `ValueTypeSet::any()`. You can imagine a future in which we know more precisely which values might be returned, given what we know about the entity and tx!

At the end of algebrizing, we know one more thing: what types a variable _ends up as_, and where else it is used.

We need to generate a constraint on all columns for which the first type set is broader than the second.

(Sometimes this might be a little redundant, but not often.)

In the unit-type case — where we later find out that a variable is bound to values of a single type — that'll take care of cross-linking, too. Otherwise, we'll restrict the primary and cross-link to the secondaries.

I don't see how you can implement this ticket without tracking the output types of each pattern as you go.